### ISSUE TITLE
bug: devcontainer is missing GDAL

### ISSUE THREAD

**Issue by anentropic** [ibis-project/ibis#9011]
### What happened?

I checked out the repo and opened it in VS Code. It found a `.devcontainer` config and prompted me to open in a container... I thought this might be easier than messing around with conda etc

But it doesn't work. The `updateContent.sh` script fails trying to pip install dependencies with an error:

```
Failed to get options via gdal-config: [Errno 2] No such file or directory: 'gdal-config'
      CRITICAL:root:A GDAL API version must be specified. Provide a path to gdal-config using a GDAL_CONFIG environment variable or use a GDAL_VERSION environment variable.
```
Presumably gdal is not installed in the `mcr.microsoft.com/vscode/devcontainers/python:3.11` base image

By adding `"remoteUser": "vscode"` to the devcontainer.json I was able to `sudo apt-get install libgdal-dev`

This got further but hit another problem:
```
Defaulting to user installation because normal site-packages is not writeable
Obtaining file:///workspaces/ibis
  Installing build dependencies ... done
  Checking if build backend supports build_editable ... done
  Getting requirements to build editable ... done
  Preparing editable metadata (pyproject.toml) ... error
  error: subprocess-exited-with-error
  
  × Preparing editable metadata (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [36 lines of output]
      Traceback (most recent call last):
        File "/usr/local/lib/python3.11/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 353, in <module>
          main()
        File "/usr/local/lib/python3.11/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 335, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/usr/local/lib/python3.11/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 181, in prepare_metadata_for_build_editable
          return hook(metadata_directory, config_settings)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/tmp/pip-build-env-qu9btf27/overlay/lib/python3.11/site-packages/poetry/core/masonry/api.py", line 42, in prepare_metadata_for_build_wheel
          poetry = Factory().create_poetry(Path(".").resolve(), with_groups=False)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/tmp/pip-build-env-qu9btf27/overlay/lib/python3.11/site-packages/poetry_dynamic_versioning/patch.py", line 26, in alt_poetry_create
          name = _get_and_apply_version(
                 ^^^^^^^^^^^^^^^^^^^^^^^
        File "/tmp/pip-build-env-qu9btf27/overlay/lib/python3.11/site-packages/poetry_dynamic_versioning/__init__.py", line 608, in _get_and_apply_version
          version, instance = _get_version(config, name)
                              ^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/tmp/pip-build-env-qu9btf27/overlay/lib/python3.11/site-packages/poetry_dynamic_versioning/__init__.py", line 452, in _get_version
          version = _get_version_from_dunamai(vcs, pattern, config)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/tmp/pip-build-env-qu9btf27/overlay/lib/python3.11/site-packages/poetry_dynamic_versioning/__init__.py", line 417, in _get_version_from_dunamai
          return Version.from_vcs(
                 ^^^^^^^^^^^^^^^^^
        File "/tmp/pip-build-env-qu9btf27/overlay/lib/python3.11/site-packages/dunamai/__init__.py", line 1956, in from_vcs
          return cls._do_vcs_callback(
                 ^^^^^^^^^^^^^^^^^^^^^
        File "/tmp/pip-build-env-qu9btf27/overlay/lib/python3.11/site-packages/dunamai/__init__.py", line 1995, in _do_vcs_callback
          return callback(**kwargs)
                 ^^^^^^^^^^^^^^^^^^
        File "/tmp/pip-build-env-qu9btf27/overlay/lib/python3.11/site-packages/dunamai/__init__.py", line 1916, in from_any_vcs
          vcs = _detect_vcs(None, path)
                ^^^^^^^^^^^^^^^^^^^^^^^
        File "/tmp/pip-build-env-qu9btf27/overlay/lib/python3.11/site-packages/dunamai/__init__.py", line 375, in _detect_vcs
          raise RuntimeError(dubious_ownership_error)
      RuntimeError: Detected Git repository, but failed because of dubious ownership
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.
```
(this is running from the VS Code terminal window after opening the project as/in a devcontainer)

### What version of ibis are you using?

git HEAD [`4ad734d`](https://github.com/ibis-project/ibis/commit/4ad734d21baca1cd53d7fa973eafb6959760d6f2)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by ncclementi** [comment]
@anentropic Thanks for reporting this. 

I'm not quite sure what's going on, because the geospatial dependencies were added recently, see https://github.com/ibis-project/ibis/pull/8837 and it looks like you are running on main. 

@cpcloud any idea what could be going on here?

**Comment by anentropic** [comment]
I believe it's running a docker container locally on my machine

from the terminal I get:
```
vscode ➜ /workspaces/ibis (main) $ arch
aarch64
vscode ➜ /workspaces/ibis (main) $ uname -a
Linux b66611250c8b 6.6.22-linuxkit #1 SMP Fri Mar 29 12:21:27 UTC 2024 aarch64 GNU/Linux
```

which makes sense as I'm on an M2 macbook

**Comment by anentropic** [comment]
for kicks I pasted in the traceback above asked the new Llama 3 for a solution...

it actually gave a nice explanation and something that worked (!)

amending the update content script to:
```sh
POETRY_DYNAMIC_VERSIONING=false python3 -m pip install -e '.[clickhouse,duckdb,clickhouse,examples,geospatial]'
```

got around the dunamai issue

So the two parts of the fix are probably:

1. add a Dockerfile under `.devcontainer` something like:

```Dockerfile
FROM mcr.microsoft.com/vscode/devcontainers/python:3.11

RUN apt-get update && apt-get install -y --no-install-recommends libgdal-dev
```

2. prepend `POETRY_DYNAMIC_VERSIONING=false` to the `pip install -e`

**Comment by gforsyth** [comment]
Hey @anentropic -- want to give it another shot now and see if it works?

**Comment by amoeba** [comment]
Hi @gforsyth, I had to make a few tweaks to get the devcontainer to run and those changes are in https://github.com/ibis-project/ibis/pull/9019.

**Comment by anentropic** [comment]
> Hey @anentropic -- want to give it another shot now and see if it works?

Not working yet but looks like #9019 will fix it (needs the change in `devcontainer.json` to actually use the Dockerfile)

**Comment by cpcloud** [comment]
Just merged that, give it another go if you can!

**Comment by anentropic** [comment]
Apologies, but it seems `POETRY_DYNAMIC_VERSIONING=false` was a hallucination by the LLM

I tried it and it appeared to work (i.e. by entering that line manually in terminal of a failed container launch)

but it doesn't work when (re)building the container from scratch ...I tried again just now and same thing. (And I can't find evidence for it via Google or GitHub search)

I'm not sure what devcontainers is doing to make the two scenarios behave differently

According to poetry-dynamic-versioning plugin docs a correct env var to disable active versioning would be `POETRY_DYNAMIC_VERSIONING_BYPASS='8.0.0'` i.e. specifying a static version value

I tried this and took care to make devcontainers rebuild the image with it and it seems to work
